### PR TITLE
feat(Algebra/GroupWithZero/Divisibility): add ad-hoc ring division

### PR DESCRIPTION
**Pull request by Multramate**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/GroupWithZero/Divisibility.lean [hunk 1]
```diff
@@ -141,3 +143,66 @@ theorem eq_of_forall_dvd' (h : ∀ c, c ∣ a ↔ c ∣ b) : a = b :=
...
+variable {M₀ : Type u} [CommMonoidWithZero M₀]
+
+lemma divide_mul_cancel {x y : M₀} (hy : y ≠ 0) (hx : y ∣ x) : divide x y * y = x := by
+  rw [mul_comm, mul_divide_cancel hy hx]
+
+lemma mul_divide_cancel_right [IsRightCancelMulZero M₀] {x y : M₀} (hy : y ≠ 0) :
+    divide (x * y) y = x := by
+  rw [divide_dvd hy <| dvd_mul_left y x]
+  exact mul_right_cancel₀ hy <| mul_comm _ y ▸ (dvd_mul_left y x).choose_spec.symm
+
```
Comment (eric-wieser) [inline Mathlib/Algebra/GroupWithZero/Divisibility.lean:207]: Can you add `divide_eq_mul_inverse` for `is_unit b`?

File: Mathlib/Algebra/GroupWithZero/Divisibility.lean [hunk 2]
```diff
@@ -141,3 +143,66 @@ theorem eq_of_forall_dvd' (h : ∀ c, c ∣ a ↔ c ∣ b) : a = b :=
...
+variable {M₀ : Type u} [MonoidWithZero M₀]
+
+/-- Introduce a binary function `divide` on monoids with zero `M₀`, which sends `x` and `y` to
+`x / y` if `y` is non-zero and divides `x`, and to `0` otherwise. This definition is somewhat
+ad hoc, but one needs a fully (rather than partially) defined division function for some purposes.
+
+Note that while this is in the `Ring` namespace for brevity, it requires the weaker assumption
+`MonoidWithZero M₀` instead of `Ring M₀`. -/
+noncomputable def divide (x y : M₀) : M₀ :=
+  if h : y ≠ 0 ∧ y ∣ x then h.right.choose else 0
```
Comment (jcommelin) [inline Mathlib/Algebra/GroupWithZero/Divisibility.lean:160]: Why do you require `y` to be non-zero? In a general ring, being non-zero isn't a very natural condition. You can still have plenty of nilpotents and/or zero-divisors.

File: Mathlib/Algebra/GroupWithZero/Divisibility.lean [hunk 3]
```diff
@@ -9,14 +9,16 @@ Neil Strickland, Aaron Anderson
...
-import Mathlib.Algebra.GroupWithZero.Basic
+import Mathlib.Algebra.GroupWithZero.Units.Basic
 import Mathlib.Algebra.Divisibility.Units
 
 /-!
 # Divisibility in groups with zero.
 
 Lemmas about divisibility in groups and monoids with zero.
 
+We also define `Ring.divide`, a globally defined function on any ring
```
Comment (jcommelin) [inline Mathlib/Algebra/GroupWithZero/Divisibility.lean:20]: Can you please elaborate on how/where you want to use this function? Would a `DivInvMonoid` work in your setting?

Comment (Multramate) [reply Mathlib/Algebra/GroupWithZero/Divisibility.lean:20]: I'm looking at polynomial rings over (say) a commutative ring, which a priori don't have division/inversion but the notion of "cancelling out factors formally" still makes sense.

File: Mathlib/Algebra/GroupWithZero/Divisibility.lean (hunk 2 above)
Comment (Multramate) [reply Mathlib/Algebra/GroupWithZero/Divisibility.lean:160]: Fair enough. In my head I'm working with only integral domains, but I just didn't want to impose this on the definition. Without this condition I wouldn't be able to prove `divide_zero`, which I think is something nice to have.

File: Mathlib/Algebra/GroupWithZero/Divisibility.lean (hunk 3 above)
Comment (jcommelin) [reply Mathlib/Algebra/GroupWithZero/Divisibility.lean:20]: Shouldn't we treat division in polynomial rings the same way we treat it in `Nat` and `Int`? So we can just use the ordinary division symbol for some "truncated division" operation.

Comment (Multramate) [reply Mathlib/Algebra/GroupWithZero/Divisibility.lean:20]: Yes I think we should, but I don't know how to best introduce this. Define `Ring.divide` as I have and then add notation?

Comment (eric-wieser) [reply Mathlib/Algebra/GroupWithZero/Divisibility.lean:20]: Isn't this what `rem` and `quot` are for? I don't remember what typeclass they are from...

Comment (Multramate) [reply Mathlib/Algebra/GroupWithZero/Divisibility.lean:20]: There are `EuclideanDomain.quotient` and `EuclideanDomain.remainder` but my polynomial rings are not Euclidean.

**Comment by Multramate** [comment]
This seems to cause a whole lot of other overloading (?) issues - any ideas?

**Comment by eric-wieser** [comment]
Making this a div instance is a very bad idea IMO

**Comment by Multramate** [comment]
Yeah, I reverted it.

File: Mathlib/Algebra/GroupWithZero/Divisibility.lean (hunk 2 above)
Comment (ericrbg) [reply Mathlib/Algebra/GroupWithZero/Divisibility.lean:160]: I mean somthing like epsX / epsX should still be well-defined, I guess?

File: Mathlib/Algebra/GroupWithZero/Divisibility.lean [hunk 4]
```diff
@@ -141,3 +143,71 @@ theorem eq_of_forall_dvd' (h : ∀ c, c ∣ a ↔ c ∣ b) : a = b :=
...
+lemma divide_dvd {x y : M₀} (hy : y ≠ 0) (hx : y ∣ x) : divide x y = hx.choose := by
+  rw [divide, dif_pos ⟨hy, hx⟩]
+
+lemma divide_not_dvd {x y : M₀} (hx : ¬y ∣ x) : divide x y = 0 := by
+  simp only [divide, dif_neg, hx, and_false]
+
+lemma divide_zero (x : M₀) : divide x 0 = 0 := by
+  simp only [divide, dif_neg, ne_eq, false_and]
+
+lemma divide_one [NeZero (1 : M₀)] (x : M₀) : divide x 1 = x := by
```
Comment (ericrbg) [inline Mathlib/Algebra/GroupWithZero/Divisibility.lean:171]: isn't this usually stated as `Nontrivial M0`?

File: Mathlib/Algebra/GroupWithZero/Divisibility.lean (hunk 2 above)
Comment (ericrbg) [reply Mathlib/Algebra/GroupWithZero/Divisibility.lean:160]: but it'd also be nice to have `zero_divide` in general. I wonder if there's another design that allows this, maybe prioritising zero if it's possible

**Review by ericrbg — COMMENTED** [review]
I think it's probably fine to leave more API building to when it's needed, `Ring.divide (a + b) x` is probably one of the obvious things but I have a feeling that's actually got stronger requirements than you'd hope.

**Comment by Multramate** [comment]
What do you mean by `epsX`? I think generalising `zero_divide` to non-domains wouldn't work (e.g. 0/2 can be 0 or 2 in Z/4Z)

(can't seem to reply directly on the review comments)

**Comment by ericrbg** [comment]
Like `epsilon` in some ring where `epsilon^2 = 0`. I meant to reply to Johan's comments so maybe that's where the reply option is.

And I get that with this definition, `zero_divide` needs a domain - I wonder if we could tweak it so that it works everywhere

**Comment by ocfnash** [comment]
I'm not convinced that we should introduce such an arbitrary function for all `MonoidWithZero`. To be convinced I think I'd need to see:

- evidence of significant benefit that this provides in some setting (e.g., your polynomial rings), and
- evidence that no other solutions exist,

(and even then I might still be dubious).

I'm going to relabel this "awaiting-author" pending such evidence.

**Comment by Multramate** [comment]
> I'm not convinced that we should introduce such an arbitrary function for all `MonoidWithZero`.
The only reason I'm defining it for a `MonoidWithZero` is to match `Ring.inverse`, which shares a similar name, but I really only need it for an integral domain as I mentioned above. I'm definitely open to restricting this.

> * evidence of significant benefit that this provides in some setting (e.g., your polynomial rings), and
I don't have my environment set up currently, but the gist of what I'm trying to do is as follows. I have a (strong odd/even) recursive function `f : \N -> R[X]` for a commutative ring `R`, such that when `n` is even, `f n` is defined to be some polynomial `g` (in terms of `f (n / 2 - 1)`, etc) divided by `f 2`. It's impossible to extract out a factor of `f 2` from `g` formally, and the only way I can think of is to use my current `Ring.divide`, and prove later that `g` is divisible by `f 2`.

> * evidence that no other solutions exist,
I don't know how to show this other than providing the example.

**Comment by Multramate** [comment]
> Can you provide such a statement (informally) or a reference with the precise details?

I'm trying to define a characteristic two version of [division polynomials](https://en.wikipedia.org/wiki/Division_polynomials), a slightly incorrect version of which can be found in Silverman's *The Arithmetic of Elliptic Curves* Exercise 3.7. The code would look something like: #6703

**Comment by ericrbg** [comment]
https://github.com/leanprover-community/mathlib4/compare/Ring.divide...chooseP

Here's an idea I had to use a `choose Preferring` (hence the name). All the theorems you proved are true (if not in more generality) and I didn't take much care to make the API nice yet, as it's just an idea. What do you think, David?

**Comment by kbuzzard** [comment]
What the application David has in mind boils down to, is the theory of elliptic divisibility sequences https://en.wikipedia.org/wiki/Elliptic_divisibility_sequence . So the question really is whether one can formalise these without having to introduce some ad hoc ring division (and I suspect we can).

**Comment by Multramate** [comment]
I am closing this after discussions with Kevin (Thanks!) :)

**Pull request closed without merging**